test/vm_tests/builder/parsing/members_test.dart

```diff
@@ -0,0 +1,1090 @@
+                .isComponent2(Version.v3_legacyDart2Only),
+            false);
+        expect(
+            BoilerplateMemberHelper.getBoilerplateMembersForVersion(BoilerplateVersions.v4)
+                .firstWhereType<BoilerplateComponent>()
+                .isComponent2(Version.v4_mixinBased),
+            true);
+      });
+
+      group('validate', () {
+        ErrorCollector collector;
+        SourceFile file;
+        var validateResults = <String>[];
+
+        void validateCallback(String message, [SourceSpan span]) {
+          validateResults.add(message);
+        }
+
+        tearDown(() {
+          validateResults = <String>[];
+          file = null;
+          collector = null;
+        });
+
+        group('does not throw when', () {
+          group('the component is a', () {
+            for (final version in BoilerplateVersions.values) {
+              test('${versionDescriptions[version]} component', () {
+                final members = BoilerplateMemberHelper.getBoilerplateMembersForVersion(version);
+                final component = members.whereType<BoilerplateComponent>().first;
+                final componentVersion = resolveVersion(members).version;
+                file = SourceFile.fromString(getBoilerplateString(version: version));
+                collector = ErrorCollector.callback(file,
+                    onError: validateCallback, onWarning: validateCallback);
+
+                component.validate(componentVersion, collector);
+                expect(validateResults, []);
+              });
+            }
+          });
+        });
+
+        group('throws when', () {
+          test('the component is mixin based but does not extend Component2', () {
+            final members = BoilerplateMemberHelper.parseAndReturnMembers(r'''
+            UiFactory<FooProps> Foo = _$Foo;
+
+            mixin FooProps on UiProps {}
+  
+            mixin FooState on UiState {}
+  
+            class FooComponent extends UiStatefulComponent<FooProps, FooState>{}
+            ''');
+
+            final component = members.whereType<BoilerplateComponent>().first;
+            final componentVersion = resolveVersion(members).version;
+            file = SourceFile.fromString(r'''
+            UiFactory<FooProps> Foo = _$Foo;
+
+            mixin FooProps on UiProps {}
+  
+            mixin FooState on UiState {}
+  
+            class FooComponent extends UiStatefulComponent<FooProps, FooState>{}
+            ''');
+            collector = ErrorCollector.callback(file,
+                onError: validateCallback, onWarning: validateCallback);
+
+            component.validate(componentVersion, collector);
+            expect(validateResults, [contains('Must extend UiComponent2, not UiComponent.')]);
+          });
+        });
+
+        // Group that ensures that iterates over the deprecated lifecycles and all
+        // versions of the boilerplate, checking for errors where necessary.
```

Comment:
Typo (another one of mine ;) )
```suggestion
        // Group that iterates over the deprecated lifecycles and all
        // versions of the boilerplate, checking for errors where necessary.
```